### ISSUE TITLE
Optional object attributes with defaults produce incorrect error when variable default value is null

### ISSUE THREAD

**Issue by fergoid** [hashicorp/terraform#31795]
### Terraform Version

```shell
1.3.0-rc1
```

### Terraform Configuration Files

```terraform
...terraform config...
```

### Debug Output

None

### Expected Behavior

If the default value for the variable is null then I should not see an error if I do not supply any values for the variable. 

### Actual Behavior

Given this

```
variable "with_optional_attribute" {
  type = object({
    a = string                # a required attribute
    b = optional(string)      # an optional attribute
    c = optional(number, 127) # an optional attribute with a default value
  })
  default = null
}
```

Should I be able to rely on the default value of null being set for the variable?
When I have this

```
output "test" {
  value = var.with_optional_attribute == null ? "default" : "set"
}
```
I get the following error:
```
╷
│ Error: Invalid value for input variable
│ 
│   on variables.tf line 11:
│   11: variable "with_optional_attribute" {
│ 
│ The given value is not suitable for var.with_optional_attribute declared at variables.tf:11,1-35: attribute "a" is required.
╵

```

### Steps to Reproduce

terraform init

**Comment by apparentlymart** [comment]
I've added the v1.3.0 milestone to this with the intention that it means we should definitely at least _triage_ this before the v1.3.0 final release.

However, if after initial review it seems like something we can address in v1.3.1 (or so) without causing trouble for folks who already began using the feature in v1.3.0 then I would suggest we take it back out of the milestone and plan to fix in a patch release.

**Comment by fergoid** [comment]
in my testing if I explicitly set the variable to null it works but the null default does not.

**Comment by alisdair** [comment]
Thanks for reporting this! We've merged a fix which will go into the 1.3.0 release.
